Add initial Local storage implementation

bencord0 opened the pull request:
Note that all "remote" operations are currently stubbed.
A "standard" config template has been added that uses
publically readable docker container images.

This is enough to configure `lal` for local development.

    `lal configure ./configs/standard.json`

clux commented on configs/standard.json:
```diff
@@ -0,0 +1,33 @@
+    {
+      "src": "/mnt/tools",
+      "dest": "/tools",
+      "readonly": true
```
this mount isn't useful in public

clux commented on configs/standard.json:
```diff
@@ -0,0 +1,33 @@
+    {
+      "src": "cargo-cache",
+      "dest": "/home/lal/.cargo",
+      "readonly": false
```
and this is broken with my muslrust image atm, probably leave that out

clux commented on src/storage/local.rs:
```diff
@@ -0,0 +1,69 @@
+/// specific low-level use cases, these methods can be used directly.
+impl Backend for LocalBackend {
+    fn get_versions(&self, _name: &str, _loc: Option<&str>) -> LalResult<Vec<u32>> {
+    	Ok(vec![])	
```
wait, what

clux commented on src/storage/local.rs:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    fn get_latest_version(&self, _name: &str, _loc: Option<&str>) -> LalResult<u32> {
+    	Ok(0)
```
so there's only ever a version 0? then the previous vector should be `vec![0]` right?

bencord0 replied:
I'm just appeasing the type system. Without somewhere to store artifacts in `LocalBackend`, versions are meaningless.

I'm open to ideas about what we do about locally publishing artifacts in this backend.

clux replied:
I feel like if we just made lal publish put things directly in stash rather than upload somewhere, then a `LocalBackend` can just query stash and avoid HTTP altogether. May need to tweak how the traits are set up though.

clux reviewed (commented):
guess this is bypassing most of the trait methods here, so am curious; what does actually work with this storage stub?

clux commented:
ah, i didn't read your main pr comment.
i added a demo template in the main repo, haven't pulled it in here yet.

clux commented on src/core/ensure.rs:
```diff
@@ -0,0 +1,15 @@
+
+
+/// Ensure a directory exists and is empty
+pub fn ensure_dir_exists_fresh(dir: &str) -> io::Result<()> {
```
Maybe expose a function that takes a Path / PathBuf directly so you don't need to convert back and forth as well.
Good place to put this helper though!

clux commented on src/storage/local.rs:
```diff
@@ -53,10 +54,22 @@ impl Backend for LocalBackend {
+        let tar_path = format!("{}{}/{}/{}.tar.gz", prefix, name, version, name);
+        let lock_path = format!("{}{}/{}/lockfile.json", prefix, name, version);
+
+        ensure_dir_exists_fresh(config_dir().join(tar_dir).to_str().unwrap())?;
```
`.to_str().unwrap()` is probably unnecessary. Paths have a `.display()` method. But see first suggestion.

clux commented on src/storage/local.rs:
```diff
@@ -53,10 +54,22 @@ impl Backend for LocalBackend {
+        let lockfile = artifactdir.join("lockfile.json");
+
+        // prefix with environment
+        let prefix = env.map(|s| format!("env/{}/", s)).unwrap_or_else(|| "global/".into());
```
Forgot that env is still an option here due to the global join location :/
This is a very nice way of doing the ternary though.

bencord0 commented:
Updated to use `pub fn ensure_dir_exists_fresh(dir: &Path) -> io::Result<()>`.

Still need (at least) an implementation for `get_tarball_url` and `raw_download` for `lal fetch` to work.

clux commented:
Ah, but that's easy. You make `get_tarball_url` return the path of the filesystem where the file should be, and raw download just check if the file is there or fail.

clux commented on src/build.rs:
```diff
@@ -172,7 +162,7 @@ pub fn build(
     if opts.release && !modes.printonly {
         trace!("Create ARTIFACT dir");
-        ensure_dir_exists_fresh("ARTIFACT")?;
+        ensure_dir_exists_fresh(Path::new("./ARTIFACT"))?;
```
now you are spreading Path's everywhere though.

clux commented on src/core/ensure.rs:
```diff
@@ -0,0 +1,14 @@
+
+
+/// Ensure a directory exists and is empty
+pub fn ensure_dir_exists_fresh(dir: &Path) -> io::Result<()> {
```
you could parametrise this over `<T>` that is `Into: Path` to get around my above comment.

clux replied:
not that this really matters, just a nitpick

bencord0 replied:
Do you have some example syntax for this?

clux replied:
yeah, don't do this actually. don't need to start the whole `Cow` usage for such small things.

clux replied:
It's a thing usually only done in API boundaries with string and str: https://jwilm.io/blog/from-str-to-cow/

clux replied:
it's overkill here anyway

bencord0 replied:
Yeah, this isn't an API boundary, this is just code reuse, but that's kinda nice.

clux commented on src/storage/local.rs:
```diff
@@ -47,10 +47,13 @@ impl Backend for LocalBackend {
                 name: name.into(),
                 version: 0,
-                tarball: "".into()
+                tarball: full_tar_path.into_os_string().into_string().unwrap()
```
you don't need unwrap here: https://doc.rust-lang.org/std/path/struct.Path.html#method.display

bencord0 commented on src/storage/local.rs:
```diff
@@ -0,0 +1,85 @@
+        Ok(Component {
+                name: name.into(),
+                version: 0,
+                tarball: full_tar_path.into_os_string().into_string().unwrap()
```
I'm not too happy with this unwrap/crash behaviour.
We should decide what to do if components have unicode names, at the very least they should be filesystem safe. Ideally, this function should fail with a suitable `LalResult`.

clux replied:
i actually want to enforce lowercase ascii names as component names anyway, so you can really just use `display()` and never think about this

bencord0 commented on src/storage/local.rs:
```diff
@@ -0,0 +1,85 @@
+        &self,
+        name: &str,
+        _version: Option<u32>,
+        _loc: Option<&str>,
```
What's a `loc`?
Is this how we handle `env`?

clux replied:
yeah, env gets translated to loc in download.rs so could be renamed, but in storage land in means the bucket location so maybe sensible

bencord0 commented on src/publish.rs:
```diff
@@ -43,8 +43,8 @@ pub fn publish<T: Backend + ?Sized>(name: &str, backend: &T, env: Option<&str>)
           name,
           version,
-          env.unwrap_or("global"));
-    backend.upload_artifact_dir(name, version, env)?;
```
I think this is/was a bug.

We're there any circumstances that caused `env` to be `Some`?
If `env` is not specified, we should use the environment that is in the lockfile.

This now lets us modify `LocalBackend::upload_artifact_dir` so that it uploads directly to the location that the `CachedBackend` expects, and bypass `get_tarball_url` and `raw_download`.

clux replied:
We would have subjobs that would:

```sh
lal build
lal -e xenial publish {name}
```

and then if all the subjobs succeeded we'd re-publish one of them to the default bucket:

```sh
lal env reset
lal -e $(lal env) export {component}=${{BUILD_NUMBER}}
echo "Republishing $(lal env) component to default"
mkdir -p ARTIFACT
tar xzf {component}.tar.gz lockfile.json
mv {component}.tar.gz lockfile.json ARTIFACT/
lal publish {component}
```

clux replied:
Or if we only had one env, we would sometimes just republish straight from the subjob.

```sh
+ lal -e xenial publish freetype
lal::publish: Publishing freetype=202 to xenial
lal::storage::artifactory: PUT https://maven.host/CME-release/env/xenial/freetype/202/freetype.tar.gz
lal::storage::artifactory: PUT https://maven.host/CME-release/env/xenial/freetype/202/freetype.tar.gz (X-Checksum-Sha1)
lal::storage::artifactory: PUT https://maven.host/CME-release/env/xenial/freetype/202/lockfile.json
lal::storage::artifactory: PUT https://maven.host/CME-release/env/xenial/freetype/202/lockfile.json (X-Checksum-Sha1)
+ lal publish freetype
lal::publish: Publishing freetype=202 to global
lal::storage::artifactory: PUT https://maven.host/CME-release/freetype/202/freetype.tar.gz
lal::storage::artifactory: PUT https://maven.host/CME-release/freetype/202/freetype.tar.gz (X-Checksum-Sha1)
lal::storage::artifactory: PUT https://maven.host/CME-release/freetype/202/lockfile.json
lal::storage::artifactory: PUT https://maven.host/CME-release/freetype/202/lockfile.json (X-Checksum-Sha1)
```

At any rate it's important to publish to the specific bucket because `fetch` is done from the specific env bucket in the environment we are in (unless overridden), but the global one is treated as the join. It's a bit messy atm. Would benefit from a rethink.

bencord0 commented on src/storage/local.rs:
```diff
@@ -0,0 +1,85 @@
+        _version: Option<u32>,
+        _loc: Option<&str>,
+    ) -> LalResult<Component> {
+        let tar_path = format!("global/{}/0/{}.tar.gz", name, name);
```
FIXME: Don't lie about the environment. Point to the cache path location.

bencord0 commented:
Okay I've changed the publish/upload_artifact_dir logic so that artifacts are "uploaded" to the cache dir.
This has the happy side effect that `lal update` and `lal fetch` now work.

(caveat, so long as the version number is `0`, I should really fix that one too)

This essentially means that local dependency resolution now works.
```
$ cd libhello
$ lal build --release --with-version=0 --with-sha=$(git rev-parse HEAD)
$ lal publish libhello

$ cd ../hellobin
$ lal fetch
$ ls INPUT
libhello
$ lal build
$ ./OUTPUT/hello
```

clux commented:
I've cherry-picked my changes into #8 so if you drop-rebase all my changes herein you are good.

bencord0 committed:
Move ensure_dir_exists_fresh to core/ensure.rs

bencord0 committed:
Add LocalBackend to src/storage

Local publishing and updates now work.

Basic dependency resolution:

    $ cd A
    $ lal build --release --with-version=1 --with-sha=$(git rev-parse HEAD)
    $ lal publish A
    lal::publish: Publishing A=1 to xenial

    $ lal -e xenial query A
    1

    cd ../B
    $ lal fetch
    lal::fetch: Fetch xenial A 1
    $ lal build
    ...

Now release a new version

    cd ../A
    $ lal build --release --with-version=2 --with-sha=$(git rev-parse HEAD)
    $ lal publish A
    lal::publish: Publishing A=2 to xenial

    $ lal -e xenial query A
    2
    1

Update dependencies in B

    $ cd ../B
    $ lal update A --save
    lal::update: Fetch xenial A
    lal::storage::download: Last versions for A in xenial env is {1, 2}
    lal::update: Fetch xenial A=2
    lal::update: Upgrading A from 1 to 2

bencord0 committed:
Version bump to 3.8.0

bencord0 commented:
```
Local publishing and updates now work.

Basic dependency resolution:

    $ cd A
    $ lal build --release --with-version=1 --with-sha=$(git rev-parse HEAD)
    $ lal publish A
    lal::publish: Publishing A=1 to xenial

    $ lal -e xenial query A
    1

    cd ../B
    $ lal fetch
    lal::fetch: Fetch xenial A 1
    $ lal build
    ...

Now release a new version

    cd ../A
    $ lal build --release --with-version=2 --with-sha=$(git rev-parse HEAD)
    $ lal publish A
    lal::publish: Publishing A=2 to xenial

    $ lal -e xenial query A
    2
    1

Update dependencies in B

    $ cd ../B
    $ lal update A --save
    lal::update: Fetch xenial A
    lal::storage::download: Last versions for A in xenial env is {1, 2}
    lal::update: Fetch xenial A=2
    lal::update: Upgrading A from 1 to 2

```

bencord0 commented:
I'm leaving this link here so that I can clean up the error handling based on documented examples.
https://rustbyexample.com/std_misc/fs.html

clux commented on src/storage/local.rs:
```diff
@@ -0,0 +1,92 @@
+            let version = u32::from_str(filename.to_str().unwrap()).unwrap();
+            versions.push(version);
+        }
+        Ok(versions)
```
Take a look at `clean.rs`, it uses `WalkDir`. The u32 parsing is fine for now given we're the only ones fiddling with those directories, but can skip at least one unwrap by using display there.

clux commented on src/storage/local.rs:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    fn get_latest_version(&self, _name: &str, _loc: &str) -> LalResult<u32> {
+        unimplemented!()
```
get_versions -> take(1) ? :D

clux commented on src/storage/local.rs:
```diff
@@ -0,0 +1,92 @@
+impl Backend for LocalBackend {
+    fn get_versions(&self, name: &str, loc: &str) -> LalResult<Vec<u32>> {
+
+        let tar_dir = format!("cache/environments/{}/{}/", loc, name);
```
probably needs to be a path that we store in the config rather than the cache dir.

clux commented on src/storage/local.rs:
```diff
@@ -0,0 +1,92 @@
+        Ok(Component {
+            name: name.into(),
+            version: version.unwrap(),
+            location: "".into(),
```
location needs to be `storage_dir().join(env).join(version).join(name.tar).display()`

clux reviewed (commented):
Getting somewhere now :+1:

bencord0 committed:
Add examples

bencord0 committed:
localstorage: Implement get_latest_version

bencord0 committed:
localstorage: handle some unwraps

bencord0 commented:
Down to only one `unwrap()` left. I'm not sure if we can remove it without significant changes outside of the scope of this PR.

@clux unless you have any further suggestions, I'm ready to merge this.

clux commented:
Hey. Busy week. I've cherry-picked all your work over in the `localbackend-tests` branch so that I am at least in sync with you. Trying to make tests better and use something very similar to what you've added in examples.

Ideally, I'd like the trait to be fully implemented before reaching master: `raw_fetch` is still `unimplemented!` and everything is still dumping everything in the `config_dir` rather than a configurable directory.

clux commented:
I've implemented the last method and made it all work with `query` and `export` in tests. Will release it in 3.8.0, but think I have cherry picked all you did. Closing this now :)
